### PR TITLE
Data-pipeline for Processing the data either from kafka or mongodb and stores the data in to druid for reporting purpose

### PR DESCRIPTION
**Pull request by sivakarthikeyanoptit**
Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

### Type of change

Please choose appropriate options.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### How Has This Been Tested?

Please describe the tests that you ran to verify your changes in the below checkboxes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Ran Test A
- [ ] Ran Test B

**Test Configuration**:
* Software versions:
* Hardware versions:

### Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by sivakarthikeyanoptit** [comment]
@rjshrjndrn kindly review this PR

**Comment by sivakarthikeyanoptit** [comment]
Basically it will create the user, install the required packages, git clone that programs make the services to run,  it will process the data and store the data to druid

**Comment by sivakarthikeyanoptit** [comment]
@keshavprasadms could you please review this PR and merge

File: ansible/data-pipeline.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+- hosts: all
```
Comment (rjshrjndrn) [inline ansible/data-pipeline.yml:2]: change the hosts to a specific group

**Review by rjshrjndrn — CHANGES_REQUESTED** [review]

File: ansible/data-pipeline.yml (hunk 1 above)
Comment (sivakarthikeyanoptit) [reply ansible/data-pipeline.yml]: changed @rjshrjndrn

**Review by rjshrjndrn — CHANGES_REQUESTED** [review]

**Comment by rjshrjndrn** [comment]
Move all variables from vars/ to defaults/

**Comment by sivakarthikeyanoptit** [comment]
Moved all variables from vars/ to defaults/

File: ansible/roles/data-pipeline/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+- name: INSTALL THE PACKGES
+  apt:
+    name: software-properties-common
+    state: present
+
+- name: ADD THE REPOSITORY
+  shell: sudo apt-add-repository universe
+
+- name: INSTALL THE PACKGES
```
[the remark this answers is not in the document]
Comment (sivakarthikeyanoptit) [reply ansible/roles/data-pipeline/tasks/main.yml:24]: done rjshrjndrn

File: ansible/roles/data-pipeline/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    version: main
+
+- name: CREATE THE USER
+  user:
+    name: "{{ USER }}"
+    state: present
+    home: "/home/{{ USER }}"
+    shell: /bin/bash
+
+- name: UPDATE THE REPOS
```
[the remark this answers is not in the document]
Comment (sivakarthikeyanoptit) [reply ansible/roles/data-pipeline/tasks/main.yml]: done rjshrjndrn

File: ansible/roles/data-pipeline/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    state: present
+    home: "/home/{{ USER }}"
+    shell: /bin/bash
+
+- name: Ensure the locale exists
+  locale_gen:
+    name: en_US.UTF-8
+    state: present
+
+- name: set as default locale
```
Comment (rjshrjndrn) [inline ansible/roles/data-pipeline/tasks/main.yml:21]: can't we combine this with above one?

File: ansible/roles/data-pipeline/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    - openjdk-8-jdk
+    - software-properties-common
+    - python3-pip
+    - python3.6-dev
+    - python3.6-venv
+    - python-virtualenv
+    - zip
+    - unzip
+     
+- name: ADD THE REPOSITORY
```
Comment (rjshrjndrn) [inline ansible/roles/data-pipeline/tasks/main.yml:39]: do we have to add the repo before or after the package installation?
Use ansible ppa module for this.

File: ansible/roles/data-pipeline/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+  shell: sudo apt-add-repository universe
+
+- name: CHANGE THE OWNERSHIP FOR THIS {{ BASEPATH }} DIRECTORY
```
Comment (rjshrjndrn) [inline ansible/roles/data-pipeline/tasks/main.yml:42]: why the permission of 777? 0755 won't do?

File: ansible/roles/data-pipeline/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    state: directory
+    owner: "{{ USER }}"
+    group: "{{ USER }}"
+
+- name: Change user and create working directory under opt dir and install python virtual environment
+  shell: "cd {{ WORKDIR }} && virtualenv --python=python3.6 spark_venv"
+  become: yes
+  become_user: "{{ USER }}"
+
+- name: Create necessary logs folders for pipeline
```
Comment (rjshrjndrn) [inline ansible/roles/data-pipeline/tasks/main.yml:62]: create all directories using one block ansible file module.

File: ansible/roles/data-pipeline/tasks/main.yml [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+    owner: "{{ USER }}"
+    group: "{{ USER }}"
+
+- name: Creating a faust service shell file in executable mode
+  copy:
+    dest: "{{ WORKDIR }}/faust_as_service/faust.sh"
+    owner: "{{ USER }}"
+    group: "{{ USER }}"	
+    mode: 0777
+    content: |
```
Comment (rjshrjndrn) [inline ansible/roles/data-pipeline/tasks/main.yml:86]: keep all these in files/ directory, and copy it over to desination. Use loop and single block

File: ansible/data-pipeline.yml [hunk 9]
```diff
@@ -0,0 +1,6 @@
+---
+- hosts: data-pipeline
```
Comment (rjshrjndrn) [inline ansible/data-pipeline.yml:2]: If this is a new group, add it in folder private_repo/ansible/inventory...

**Review by rjshrjndrn — CHANGES_REQUESTED** [review]

File: ansible/roles/data-pipeline/tasks/main.yml (hunk 4 above)
Comment (sivakarthikeyanoptit) [reply ansible/roles/data-pipeline/tasks/main.yml:21]: both are different modules

File: ansible/roles/data-pipeline/tasks/main.yml (hunk 6 above)
Comment (sivakarthikeyanoptit) [reply ansible/roles/data-pipeline/tasks/main.yml:42]: done

File: ansible/roles/data-pipeline/tasks/main.yml (hunk 4 above)
Comment (sivakarthikeyanoptit) [reply ansible/roles/data-pipeline/tasks/main.yml:21]: locale_gen module will accept only two parameters  (name and state) 
so we may not be able to merge those two tasks

File: ansible/roles/data-pipeline/tasks/main.yml (hunk 5 above)
Comment (sivakarthikeyanoptit) [reply ansible/roles/data-pipeline/tasks/main.yml:39]: with the reference to the above comments i have added the repository before the installation package, regarding the usage the ppa module, i am not seeing universe repository under ppa module

File: ansible/roles/data-pipeline/tasks/main.yml (hunk 7 above)
Comment (sivakarthikeyanoptit) [reply ansible/roles/data-pipeline/tasks/main.yml:62]: done

File: ansible/roles/data-pipeline/tasks/main.yml (hunk 8 above)
Comment (sivakarthikeyanoptit) [reply ansible/roles/data-pipeline/tasks/main.yml:86]: for each file the copying content is different, even though we iterate file in the single loop 
we will not be able to copy the content

File: ansible/data-pipeline.yml (hunk 9 above)
Comment (sivakarthikeyanoptit) [reply ansible/data-pipeline.yml:2]: done

File: ansible/roles/data-pipeline/tasks/main.yml [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+- name: Ensure the locale exists
+  locale_gen:
+    name: en_US.UTF-8
+    state: present
+
+- name: set as default locale
+  command: locale-gen en_US.UTF-8
+
+- name: ADD THE REPOSITORY
```
Comment (rjshrjndrn) [inline ansible/roles/data-pipeline/tasks/main.yml:24]: use ansible module.

File: ansible/roles/data-pipeline/tasks/main.yml [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+- name: Change user and create working directory under opt dir and install python virtual environment
+  shell: "cd {{ WORKDIR }} && virtualenv --python=python3.6 spark_venv"
+  become: yes
+  become_user: "{{ USER }}"
+
+- name: Create necessary logs folders for pipeline
+  become_user: "{{ USER }}"
+  become: yes
+  shell: "cd {{ WORKDIR }} && mkdir -p logs/observation/status logs/observation_evidence logs/project logs/survey logs/survey_evidence"
```
Comment (rjshrjndrn) [inline ansible/roles/data-pipeline/tasks/main.yml:63]: why don't you use the above file module to do the same. add these directories in that loop

File: ansible/roles/data-pipeline/tasks/main.yml (hunk 8 above)
Comment (rjshrjndrn) [reply ansible/roles/data-pipeline/tasks/main.yml:77]: we don't care about content. That's why we're using different files. Add this to the fil copy module.

**Review by rjshrjndrn — CHANGES_REQUESTED** [review]

File: ansible/roles/data-pipeline/tasks/main.yml (hunk 10 above)
Comment (sivakarthikeyanoptit) [reply ansible/roles/data-pipeline/tasks/main.yml:24]: done

File: ansible/roles/data-pipeline/tasks/main.yml (hunk 5 above)
Comment (sivakarthikeyanoptit) [reply ansible/roles/data-pipeline/tasks/main.yml]: done

File: ansible/roles/data-pipeline/tasks/main.yml (hunk 11 above)
Comment (sivakarthikeyanoptit) [reply ansible/roles/data-pipeline/tasks/main.yml]: done

File: ansible/roles/data-pipeline/tasks/main.yml [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+    state: present
+
+- name: set as default locale
+  command: locale-gen en_US.UTF-8
+
+- name: ADD THE REPOSITORY
+  apt_repository:
+    repo: "{{ item }}"
+  loop:
+    - "deb http://archive.ubuntu.com/ubuntu/ bionic universe"
```
Comment (rjshrjndrn) [inline ansible/roles/data-pipeline/tasks/main.yml:28]: you can't have version specific repos. what if tomorrow we upgrade to 20 which is focal?

Comment (rjshrjndrn) [reply ansible/roles/data-pipeline/tasks/main.yml:28]: and why are we adding universe repo? all the packages below are available from the default ones, right?

**Review by rjshrjndrn — CHANGES_REQUESTED** [review]

File: ansible/roles/data-pipeline/tasks/main.yml (hunk 12 above)
Comment (sivakarthikeyanoptit) [reply ansible/roles/data-pipeline/tasks/main.yml:28]: we are facing issue with apt repository for universe,
That's why we went with shell command but as you suggested to use apt-repo,
i went with this approach 
This is were i refered 
https://github.com/ansible/ansible/issues/48714

File: ansible/roles/data-pipeline/tasks/main.yml [hunk 13]
```diff
@@ -0,0 +1,124 @@
+---
+# tasks file for data-pipeline
+- name: Git checkout
+  git:
+    repo: 'https://github.com/Sunbird-Ed/ml-analytics-service.git'
+    dest: "/opt/sparkjobs/ml-analytics-service"
+    version: main
```
Comment (rjshrjndrn) [inline ansible/roles/data-pipeline/tasks/main.yml:7]: Make a variable called "{{ ml_analytics_version|default('main') }}"

**Review by rjshrjndrn — DISMISSED** [review]

File: ansible/roles/ml-data-pipeline/tasks/main.yml [hunk 14]
```diff
@@ -0,0 +1,113 @@
+---
+# tasks file for data-pipeline
+- name: Git checkout
+  git:
+    repo: 'https://github.com/Sunbird-Ed/ml-analytics-service.git'
+    dest: "/opt/sparkjobs/ml-analytics-service"
+    version: "{{ ml_analytics_version|default('release-4.0.0') }}"
```
Comment (rjshrjndrn) [inline ansible/roles/ml-data-pipeline/tasks/main.yml:7]: @sivakarthikeyanoptit remove the default value. We will always specify which version to deploy. else there are chances for erroneous deployments.

File: ansible/roles/ml-data-pipeline/tasks/main.yml [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+    virtualenv: "{{WORKDIR }}/spark_venv"
+    virtualenv_python: python3.6
+    requirements: "{{ WORKDIR }}/ml-analytics-service/requirements.txt"
+
+- name: Creating a faust service shell file in executable mode
+  copy:
+    src: files/faust.sh
+    dest: "{{ WORKDIR }}/faust_as_service/faust.sh"
+    owner: "{{ USER }}"
+    group: "{{ USER }}"
```
Comment (rjshrjndrn) [inline ansible/roles/ml-data-pipeline/tasks/main.yml:75]: add `backup: yes`, in case there is any different version went in, we'll have the backup to compare to.

File: ansible/roles/ml-data-pipeline/tasks/main.yml [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+  with_items:
+    - faust_observation.service
+    - faust_observation_evidence.service
+    - faust_survey.service
+    - faust_survey_evidence.service
+
+- name: Template config.sample to config.ini
+  template:
+    src: "{{ config_path }}/config.sample"
+    dest: "{{ config_path }}/config.ini"
```
Comment (rjshrjndrn) [inline ansible/roles/ml-data-pipeline/tasks/main.yml:91]: here also add `backup: yes`

**Review by rjshrjndrn — CHANGES_REQUESTED** [review]

**Comment by rjshrjndrn** [comment]
@keshavprasadms apart from the above comments looks good to me. Please have a look.

File: ansible/roles/ml-data-pipeline/tasks/main.yml (hunk 14 above)
Comment (sivakarthikeyanoptit) [reply ansible/roles/ml-data-pipeline/tasks/main.yml]: done

File: ansible/roles/ml-data-pipeline/tasks/main.yml (hunk 15 above)
Comment (sivakarthikeyanoptit) [reply ansible/roles/ml-data-pipeline/tasks/main.yml:75]: done

File: ansible/roles/ml-data-pipeline/tasks/main.yml (hunk 16 above)
Comment (sivakarthikeyanoptit) [reply ansible/roles/ml-data-pipeline/tasks/main.yml:92]: done

File: pipelines/deploy/ml-analytics-service/Jenkinsfile [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+                    currentWs = sh(returnStdout: true, script: 'pwd').trim()
+                    envDir = sh(returnStdout: true, script: "echo $JOB_NAME").split('/')[-3].trim()
+                    module = sh(returnStdout: true, script: "echo $JOB_NAME").split('/')[-2].trim()
+                    jobName = sh(returnStdout: true, script: "echo $JOB_NAME").split('/')[-1].trim()
+                    values.put('currentWs', currentWs)
+                    values.put('env', envDir)
+                    values.put('module', module)
+                    values.put('jobName', jobName)
+                    ansiblePlaybook = "${currentWs}/ansible/ml-data-pipeline.yml"
+                    ansibleExtraArgs = "--vault-password-file /var/lib/jenkins/secrets/vault-pass"
```
Comment (rjshrjndrn) [inline pipelines/deploy/ml-analytics-service/Jenkinsfile:30]: you have to pass the variable value for `ml_analytics_version`
`ansibleExtraArgs = "-e ml_analytics_version=${params.ml_analytics_version} --vault-password-file /var/lib/jenkins/secrets/vault-pass"`

And in jenkins create a string parameter with variable name `ml_analytics_version`

**Review by rjshrjndrn — CHANGES_REQUESTED** [review]

File: pipelines/deploy/ml-analytics-service/Jenkinsfile (hunk 17 above)
Comment (sivakarthikeyanoptit) [reply pipelines/deploy/ml-analytics-service/Jenkinsfile]: done

File: ansible/ml-data-pipeline.yml [hunk 18]
```diff
@@ -0,0 +1,6 @@
+---
+- hosts: ml-data-pipeline
+  become: yes
+  roles:
+    - role: ml-data-pipeline
+      BASEPATH: "/opt/sparkjobs" # Git Clone Path
```
Comment (keshavprasadms) [inline ansible/ml-data-pipeline.yml:6]: Sorry what is this? I don't recall anything ansible inbuilt function in this name. If its a variable, please add under the `vars:` section.

File: ansible/roles/ml-data-pipeline/tasks/main.yml [hunk 19]
```diff
@@ -0,0 +1,115 @@
+---
+# tasks file for data-pipeline
+- name: Git checkout
+  git:
+    repo: 'https://github.com/Sunbird-Ed/ml-analytics-service.git'
+    dest: "/opt/sparkjobs/ml-analytics-service"
```
Comment (keshavprasadms) [inline ansible/roles/ml-data-pipeline/tasks/main.yml:6]: Can we use the `BASEPATH` variable here?

File: ansible/roles/ml-data-pipeline/tasks/main.yml [hunk 20]
```diff
@@ -0,0 +1,115 @@
+---
+# tasks file for data-pipeline
+- name: Git checkout
+  git:
```
Comment (keshavprasadms) [inline ansible/roles/ml-data-pipeline/tasks/main.yml:4]: Will this module re-clone or update the branch / tag on a second run? Since I dont see any cleanup of the git repo at he end of the task.

File: ansible/roles/ml-data-pipeline/tasks/main.yml [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+    state: present
+    home: "/home/{{ USER }}"
+    shell: /bin/bash
+
+- name: INSTALL THE PACKGES
+  apt:  
+    update_cache: yes
+    name: "{{ item }}"
+    state: present
+  loop:
```
Comment (keshavprasadms) [inline ansible/roles/ml-data-pipeline/tasks/main.yml:21]: No looping for apt. Its deprecated since ansible 2.7. Use something like this - https://github.com/project-sunbird/sunbird-devops/blob/master/ansible/provision.yml#L47

File: ansible/roles/ml-data-pipeline/tasks/main.yml [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+    path: "{{ item }}"
+    mode: "0755"
+    recurse: yes
+    state: directory
+    owner: "{{ USER }}"
+    group: "{{ USER }}"
+  loop:
+    - "{{ BASEPATH }}"
+    - "{{ WORKDIR }}"
+    - "{{ WORKDIR }}/faust_as_service"
```
Comment (keshavprasadms) [inline ansible/roles/ml-data-pipeline/tasks/main.yml:42]: If we doing a recurisive ownership change, why is this line required? This will already get covered from the previous line right?

File: ansible/roles/ml-data-pipeline/tasks/main.yml [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+    group: "{{ USER }}"
+    mode: "0755"
+  loop: 
+    - "{{ WORKDIR }}/logs/observation/status"
+    - "{{ WORKDIR }}/logs/observation_evidence"
+    - "{{ WORKDIR }}/logs/project"
+    - "{{ WORKDIR }}/logs/survey"
+    - "{{ WORKDIR }}/logs/survey_evidence"
+
+- name: Initiate virtualenv
```
Comment (keshavprasadms) [inline ansible/roles/ml-data-pipeline/tasks/main.yml:64]: This task name is confusing. Are we install venv or requirements here? Venv is already installed in previous steps rigjt?

File: ansible/roles/ml-data-pipeline/tasks/main.yml [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+    enabled: yes
+    state: started
+  loop:
+    - faust_observation.service 
+    - faust_observation_evidence.service 
+    - faust_survey.service 
+    - faust_survey_evidence.service
+
+- name: CREATE THE CRON
+  shell: |
```
Comment (keshavprasadms) [inline ansible/roles/ml-data-pipeline/tasks/main.yml:111]: Please use the `ansible.builtin.cron` module for this. This might overwrite or keep appending same cron on multiple runs.

File: private_repo/ansible/inventory/dev/Core/hosts [hunk 25]
```diff
@@ -197,6 +197,9 @@ yarn
 core
 local
 
+[ml-data-pipeline]
```
Comment (keshavprasadms) [inline private_repo/ansible/inventory/dev/Core/hosts:200]: Why do we need this in Core directory?

**Review by keshavprasadms — CHANGES_REQUESTED** [review]
Please add the jenkins job config.xml also with the PR so it can reviewed.

File: ansible/roles/ml-data-pipeline/tasks/main.yml (hunk 21 above)
Comment (rjshrjndrn) [reply ansible/roles/ml-data-pipeline/tasks/main.yml:21]: else give it as 
```
name:
  - package1
  - package2
  - ...
```

File: private_repo/ansible/inventory/dev/Core/hosts (hunk 25 above)
Comment (rjshrjndrn) [reply private_repo/ansible/inventory/dev/Core/hosts:200]: By the next release, it'll be containerized and moved to core. If we keep this is DP now, we'll have to migrate it from there.

File: ansible/roles/ml-data-pipeline/tasks/main.yml (hunk 22 above)
Comment (sivakarthikeyanoptit) [reply ansible/roles/ml-data-pipeline/tasks/main.yml:42]: faust_as_service directory has been created here only so we can not able to remove this part

File: ansible/roles/ml-data-pipeline/tasks/main.yml (hunk 20 above)
Comment (sivakarthikeyanoptit) [reply ansible/roles/ml-data-pipeline/tasks/main.yml:4]: done

File: ansible/ml-data-pipeline.yml (hunk 18 above)
Comment (sivakarthikeyanoptit) [reply ansible/ml-data-pipeline.yml]: i have removed it because it is mentioned in defaults

File: ansible/roles/ml-data-pipeline/tasks/main.yml (hunk 19 above)
Comment (sivakarthikeyanoptit) [reply ansible/roles/ml-data-pipeline/tasks/main.yml]: defined the basepath as per suggestion

File: ansible/roles/ml-data-pipeline/tasks/main.yml (hunk 21 above)
Comment (sivakarthikeyanoptit) [reply ansible/roles/ml-data-pipeline/tasks/main.yml]: done

File: ansible/roles/ml-data-pipeline/tasks/main.yml (hunk 23 above)
Comment (sivakarthikeyanoptit) [reply ansible/roles/ml-data-pipeline/tasks/main.yml:64]: PIP INSTALLATION OF NECESSARY PACKAGE FROM requirement.txt

File: ansible/roles/ml-data-pipeline/tasks/main.yml (hunk 24 above)
Comment (sivakarthikeyanoptit) [reply ansible/roles/ml-data-pipeline/tasks/main.yml:111]: done

File: ansible/roles/data-pipeline/tasks/main.yml (hunk 13 above)
Comment (sivakarthikeyanoptit) [reply ansible/roles/data-pipeline/tasks/main.yml]: done

File: ansible/roles/ml-data-pipeline/tasks/main.yml (hunk 24 above)
Comment (keshavprasadms) [reply ansible/roles/ml-data-pipeline/tasks/main.yml:111]: Please dont resolve conversation without fixing code.

File: private_repo/ansible/inventory/dev/DataPipeline/hosts [hunk 26]
```diff
@@ -61,6 +61,9 @@ cassandra
 [influxdb]
 18.3.0.4
 
+[ml-data-pipeline]
```
Comment (keshavprasadms) [inline private_repo/ansible/inventory/dev/DataPipeline/hosts:64]: Remove this

File: ansible/roles/ml-data-pipeline/tasks/main.yml (hunk 23 above)
Comment (keshavprasadms) [reply ansible/roles/ml-data-pipeline/tasks/main.yml:64]: I am not sure if you understand by what renaming means. Change to something meaningful please!
`name: Install packages from requirement.txt`
Don't resolve comments without making the change!

**Comment by keshavprasadms** [comment]
Where is the Jenkins job config.xml?

File: private_repo/ansible/inventory/dev/DataPipeline/hosts (hunk 26 above)
Comment (sivakarthikeyanoptit) [reply private_repo/ansible/inventory/dev/DataPipeline/hosts]: removed

**Review by keshavprasadms — APPROVED** [review]
LGTM

File: deploy/jenkins/jobs/Deploy/jobs/dev/jobs/Core/jobs/ml-analytics-service/config.xml [hunk 27]
```diff
@@ -0,0 +1,133 @@
...
+    <org.jenkinsci.plugins.workflow.job.properties.PipelineTriggersJobProperty>
+      <triggers/>
+    </org.jenkinsci.plugins.workflow.job.properties.PipelineTriggersJobProperty>
+  </properties>
+  <definition class="org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinition" plugin="workflow-cps@2.80">
+    <scm class="hudson.plugins.git.GitSCM" plugin="git@4.2.2">
+      <configVersion>2</configVersion>
+      <userRemoteConfigs>
+        <hudson.plugins.git.UserRemoteConfig>
+          <url>https://github.com/sivakarthikeyanoptit/sunbird-devops.git</url>
```
Comment (rjshrjndrn) [inline deploy/jenkins/jobs/Deploy/jobs/dev/jobs/Core/jobs/ml-analytics-service/config.xml:108]: @sivakarthikeyanoptit Please change the repo to project-sunbird

File: deploy/jenkins/jobs/Deploy/jobs/dev/jobs/Core/jobs/ml-analytics-service/config.xml [hunk 28]
```diff
@@ -0,0 +1,133 @@
...
+        </hudson.plugins.git.UserRemoteConfig>
+      </userRemoteConfigs>
+      <branches>
+        <hudson.plugins.git.BranchSpec>
+          <name>data-pipeline-3.9.0</name>
```
Comment (rjshrjndrn) [inline deploy/jenkins/jobs/Deploy/jobs/dev/jobs/Core/jobs/ml-analytics-service/config.xml:113]: branch should be ${branch_or_tag}

File: deploy/jenkins/jobs/Deploy/jobs/dev/jobs/Core/jobs/ml-analytics-service/config.xml [hunk 29]
```diff
@@ -0,0 +1,133 @@
...
+        </hudson.plugins.git.BranchSpec>
+      </branches>
+      <doGenerateSubmoduleConfigurations>false</doGenerateSubmoduleConfigurations>
+      <submoduleCfg class="list"/>
```
Comment (rjshrjndrn) [inline deploy/jenkins/jobs/Deploy/jobs/dev/jobs/Core/jobs/ml-analytics-service/config.xml:117]: where are you defining `ml_analytics_version` which is the argument for the repo clone. It's defined in the Jenkinsfile below, but not in this config. Please add it to the Jenkins job, and create a new config.xml.

File: deploy/jenkins/jobs/Deploy/jobs/dev/jobs/Core/jobs/ml-analytics-service/config.xml [hunk 30]
```diff
@@ -0,0 +1,133 @@
+<?xml version='1.1' encoding='UTF-8'?>
```
Comment (rjshrjndrn) [inline deploy/jenkins/jobs/Deploy/jobs/dev/jobs/Core/jobs/ml-analytics-service/config.xml:1]: remove the above vim swap file.

**Review by rjshrjndrn — CHANGES_REQUESTED** [review]

File: deploy/jenkins/jobs/Deploy/jobs/dev/jobs/Core/jobs/ml-analytics-service/config.xml (hunk 28 above)
Comment (sivakarthikeyanoptit) [reply deploy/jenkins/jobs/Deploy/jobs/dev/jobs/Core/jobs/ml-analytics-service/config.xml]: done

File: deploy/jenkins/jobs/Deploy/jobs/dev/jobs/Core/jobs/ml-analytics-service/config.xml (hunk 27 above)
Comment (sivakarthikeyanoptit) [reply deploy/jenkins/jobs/Deploy/jobs/dev/jobs/Core/jobs/ml-analytics-service/config.xml]: done

Comment (sivakarthikeyanoptit) [reply deploy/jenkins/jobs/Deploy/jobs/dev/jobs/Core/jobs/ml-analytics-service/config.xml]: done

File: deploy/jenkins/jobs/Deploy/jobs/dev/jobs/Core/jobs/ml-analytics-service/config.xml (hunk 29 above)
Comment (sivakarthikeyanoptit) [reply deploy/jenkins/jobs/Deploy/jobs/dev/jobs/Core/jobs/ml-analytics-service/config.xml:116]: done

**Review by rjshrjndrn — APPROVED** [review]
